### ISSUE TITLE
Type Predicate Function Inconsistent Across Identical Inferred Types

### ISSUE THREAD

**Issue by joealden** [microsoft/TypeScript#49514]
# Bug Report

### 🔎 Search Terms

type predicate filter flatMap null

### 🕗 Version & Regression Information

This is the behavior in every version I tried (that don't throw errors), and I reviewed all FAQ entries.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?target=99&ts=4.7.2#code/JYWwDg9gTgLgBAbzgMTgXzgMyhEcBEMAzgLQwQQA2ARgKaUz4DcAUCwMYQB2R8RAhjGBFMwWkTgBeFnDgAeACoB1YABNaAGgB8ACgCUUrTPkKAcvyg4A7nFoAPGLS6qJytbV1cL1gFwoAdOaWEFaKQdZaBpJGsrJewVasLAD0AFRSkplZ2Tm5efkFhUV5cKnJbJw88MKmEDCmAK6UlFIm2joAbvyUDbR+CnAAPnBcTZR6fl09tHDCcAPRcFO9cACEmSNjSTAAnmAzAJKOeJJwAIKW-DtySMt9cLxQwFwA5kObzehaSZW8s8cARlaAiEIjERDkR1oIF0eh0AG1bt1en58AD8OgALp6Vi-arHABMwMEwlE4khx1hCKR01RBIxaGxuO4f2AxwAzMTQWSIVCYfpqUtkfdRp9GTiKiz4HciEDTvC2dCARp-tCCSrFSB2Zj-JhKIIALL8MA6TWGVUgfwgY06MDmsD+O56CV4rDABi0KC0VQANWFstaMoBuvdjigpqItXqYxxcGSyQeQk+lRg-GeEgABqLKBnZlw4Lt9nAfDpJVUhdMiAS-BcoFc5GiMcN8PT3tmtK14Y2VS38Crs5imHGEwIQDMvZBYN6C3sZvxMzKMxwpW6PV7ff6iacZQSQ2uI1HGs1Y-G4FmxrnOJZaOwYJQdnAXrRiLY7OweupVFgcHhC7QWEAA)

### 💻 Code

```ts
import { F } from "ts-toolbelt"; // can easily be replaced with raw TS code if needed

const satisfies =
  <TWide,>() =>
  <TNarrow extends TWide>(narrow: F.Narrow<TNarrow>) =>
    narrow;

const isNotNull = <T,>(value: T | null): value is T => value !== null;

type Item = Array<{ value: string | null }>;

const item1 = satisfies<Item>()([{ value: "1" }]);
const item2 = satisfies<Item>()([{ value: "2" }]);
const item3 = satisfies<Item>()([{ value: null }]);

const values1 = [item1, item2, item3].flatMap(item => item.map(p => p.value));
const filteredValues1 = values1.filter(isNotNull); // still contains `null` in type :(

const values2: Array<"1" | "2" | null> = ["1", "2", null]; // same reported type as `values`
const filteredValues2 = values2.filter(isNotNull); // `null` correctly gets excluded from type
```

### 🙁 Actual behavior

`filteredValues1` still contains `null` in its type when is shouldn't.

### 🙂 Expected behavior

`filteredValues1` should have the same type as `filteredValues2` (without `null`).

---

Originated from this Twitter thread: https://twitter.com/joealden_/status/1536335566891061249

**Comment by RyanCavanaugh** [comment]
A repro with an inlined `F` would be appreciated if anyone has time

**Comment by joealden** [comment]
> A repro with an inlined `F` would be appreciated if anyone has time

@RyanCavanaugh here you go:

[Playground link with relevant code](https://www.typescriptlang.org/play?target=99&ts=4.7.2#code/C4TwDgpgBAKgTiAPAQQIxQgD2BAdgEwGcoBDXEAGimQCYNs8jTyqBhE4AYwAsoBeKLggA3CHAB8-auiw4CxWlAD80qAC4o7LtwDcAKD2hIUAHIk4cAPYB3EgCMANtAGFgcAJa4A5lAA+ggFcAWzsxPyg7dy9PYHC7S0snMn1DcGgzCxsAJRJrFEkBAApkejkmAG0AXWUoKvVBETEASnDi0sZiDKtbR2gVEo0hUTgW-0KAbz0oadqAaShPKABrCBBLADNqSo1kctnq2Q6oADEA3E5gd0tcGt39+q7s3JQ9yvF9AF8mlKN0826UO15MwQAVYAgUFQqlRHtYcnlkOJ3gZONdXFBCBx3IR1u4IMQ+FMoIgYAB1dz4CAUcSFFp8cREkmwoFMMkUiA03D-GwaWFM7nWcR0hkzQQClIAegAVPw+HL5QrFUrlSrVWr1cqoFKJSi0bFsSZLMATAEHA4pCTqYVhCQHAEIBoYOFcKaHE0NDa7dBsbB+JJPfaoABCOWBM0-NJQACSOCCUmQFhISHGUADDoxbk8Pn8LrNUA+yL0qNw6PcsfQLixOLxhEQMYgQRpTUK5RTaY0ACJUB385VvkW9QtY3RK5dq-i67Gmy227b7Z2aD2Pn39MXS7GAMxSTFj3ET+uN2kz1Nz9O583L-sDkuxNOECu1MsN1BUJ9BGivzeVAB06wcHAAWRIMBCjfP0hwbb8gmAwowHAsBvzTJp+zXWJcQcHA4AgfAADVT3vKQ71QX93AwsRQMIQ1jVdb4oAlCUM1I81i2AEhPGIAADc8OIWG5fnUQpdRvE8vUIGgdkTJAux7fwO0XZ1XTBcppKoOSOyoc9Kh0OiGMxIJoCwsBLDgHB8CgfiSE4u8OOvdF0Mw7C8NEkcRPtMSSLIuAKKok0zVo+ioC410eNRCwIAuBwQCgLwIGAYgsE4O1KTM9YrDjX49CAA)

```ts
type Try<A1 extends any, A2 extends any, Catch = never> = A1 extends A2 ? A1 : Catch;

type Narrowable = string | number | bigint | boolean;

type NarrowRaw<A> = (A extends [] ? [] : never) | (A extends Narrowable ? A : never) | ({
    [K in keyof A]: A[K] extends Function ? A[K] : NarrowRaw<A[K]>;
});

type Narrow<A extends any> = Try<A, [], NarrowRaw<A>>;

const satisfies =
  <TWide,>() =>
  <TNarrow extends TWide>(narrow: Narrow<TNarrow>) =>
    narrow;

/* ========================================================================== */

const isNotNull = <T,>(value: T | null): value is T => value !== null;

type Item = Array<{ value: string | null }>;

const item1 = satisfies<Item>()([{ value: "1" }]);
const item2 = satisfies<Item>()([{ value: "2" }]);
const item3 = satisfies<Item>()([{ value: null }]);

const values1 = [item1, item2, item3].flatMap(item => item.map(p => p.value));
const filteredValues1 = values1.filter(isNotNull); // still contains `null` in type :(

const values2: Array<"1" | "2" | null> = ["1", "2", null]; // same reported type as `values`
const filteredValues2 = values2.filter(isNotNull); // `null` correctly gets excluded from type

```

**Comment by JoanVicens** [comment]
I'm having the same problem with the filter function filtering an array of any.
What seems odd is that the same line inside a function works fine.

[Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAcg9sGBXANiqBeKBnYAnASwDsBzKAHyiKQFsAjCPCqOuOFCAQyOexHvbM4dAFYQAxsADcAKBkAzJEUkE4PAtniJUKABQA3TiiQQAXFG4gAlOcPHoG2AmRooAbxlQoeCMCR4eOxMoAEIMLGo0WQBfOUVlYFUeeQIUYEYXFGwDIxNscwBBPDxOEAAeSwA+GygikvKtTMr3T29ff0DciGwAOhS0xl0NRp0rGLlxNVwoIO7MKABtACJdKEBRRigrJYAaKABWADZdyJRdpQATCBSiCHPd-BNdtyJOGjMoJfkl6IBdWRlJkRpkRnDp5v10nhMtlZtgxlAAPQItp+ALYKCgSDmEZoBY-AFTYBQcTcABCEEy81hfVSkKGmlBaHhSJRHXRmPeulwhFIzGo9EYzFY7C4PEobior3e3OIJCkUGifLBlAuV2ItyseIJQKJAH0qV1ehDGGUcShKjl7HD-vEVGooMaoTpsAB1AjAAAWcCQwAASu0Apa8oViqUKkQQNVbHACOcWl5AcDGVkDVaaQM8PSzczkYkwHAnNpcfjYkA)


```ts
type NotNull = string | number | boolean | symbol | object;

function isNotNull(value: any): value is NotNull {
  return value !== null;
}

function filterNulls(values: Array<any>): Array<NotNull> {
  return values.filter(isNotNull);
}

const values = ["( ㅅ )", 56, null, undefined, true, {name: "f"}];

const notNull = filterNulls(values); // returns type: NotNull[]
const canBeNull = values.filter(isNotNull); // returns type: (string | number | boolean | { name: string; } | null | undefined)[]
const _ = values.filter<NotNull>(values); // Type 'NotNull' does not satisfy the constraint 'string | number | boolean | { name: string; } | null | undefined'.

function filterNullsWithoutReturn(values: Array<any>): void {
  const notNulls = values.filter(isNotNull); // returns type: NotNull[]
}
`

**Comment by andrewbranch** [comment]
Well, after a couple hours debugging, I know what’s causing this, but I’m not yet sure what to do about it. In the working `values2` case, the `"1" | "2"` constituents of the union type are “fresh literal” types. In `values1`, they are not fresh, and so they get widened to `string` when we are instantiating the signature of `isNotNull`. (The inference candidate for its `T` is the union of literals, but at the last possible moment in `getCovariantInference`, it sees no reason not to widen.) Once we have inferred `isNotNull<string>`, we can no longer choose the type predicate overload for `.filter` because `string` is not a subtype of `"1" | "2" | null`, so the non-narrowing signature is chosen.

Literal freshness is not a system I’m super familiar with in the compiler. Can anyone advise on how this is expected to work? (@weswigham @ahejlsberg?)

**Comment by ahejlsberg** [comment]
Here's a shorter example to illustrate the surprising effects of preserving literal freshness in array literal elements and object literal properties:

```ts
declare function f1<T extends "1" | "2" | "3">(x: T): T;

const value = f1("1");  // regular "1"
let x1 = value;  // regular "1"

declare function f2<T extends "1" | "2" | "3">(x: { value: T }): { value: T };

const obj2 = f2({ value: "1" });  // { value: regular "1" }
let x2 = obj2.value;  // regular "1"

declare function f3<T extends { value: "1" | "2" | "3" }>(obj: T): T;

const obj3 = f3({ value: "1" });  // { value: fresh "1" }
let x3 = obj3.value;  // string
```

The `obj3` example definitely doesn't seem right, so I think it makes sense to always infer regular (non-fresh) literal types for array literal elements and object literal properties.

**Comment by david10sing** [comment]
Hi @andrewbranch @ahejlsberg 

I've tracked this down as the source of a `ts(2345)` error in a case documented in this closed issue https://github.com/microsoft/TypeScript/issues/52125

We can get around it by using `as` but are we going to have to do this for everything now? so that the end type is `string[]`

Maybe we should update the `array.includes` definition to have `searchElement` as `unknown` instead of `T` which would make more sense because the `searchElement` could not be included in the array; hence the whole point of carrying out the `includes` check.

Looking forward to an answer to this.
